File: c/concatenation-of-consecutive-binary-numbers.rs
const MOD : i64 = 1_000_000_007;

impl Solution {
  pub fn concatenated_binary(n: i32) -> i32 {
    let mut ans = 0 as i64;
    for i in 1..=n as i64 {
      let mut pw = 1;
      for d in 0..40 {
        if pw * 2 > i { break; }
        pw *= 2;
      }
      ans = (ans * pw * 2 + i as i64) % MOD;
    }
    ans as i32
  }
}

File: c/concatenation-of-consecutive-binary-numbers.scala
object Solution {
    def concatenatedBinary(n: Int): Int = {
        val MOD = 1_000_000_007L
        var (ans, p, i) = (0L, 1L, 1L)
        while (i <= n) {
            if ((i & (i - 1)) == 0) p = (p * 2) % MOD
            ans = ((ans * p) + i) % MOD
            i += 1
        }
        ans.toInt
    }
}
